packages/framework/tests/Unit/Model/Order/Processing/OrderProcessorMiddleware/AddProductsMiddlewareTest.php

```diff
@@ -0,0 +1,165 @@
+            $productData->name = ['en' => $productProviderInput['name']];
+            $productData->vatsIndexedByDomainId = [1 => $this->createVat()];
+            $productData->unit = $this->createMock(Unit::class);
+            $product = Product::create($productData);
+
+            $orderProcessingData->orderInput->addProduct($product, $productProviderInput['quantity']);
+            $products[] = $product;
+            $quantifiedItemPrices[] = new QuantifiedItemPrice(
+                $productProviderInput['unitPrice'],
+                $productProviderInput['unitPrice']->multiply($productProviderInput['quantity']),
+                $this->createVat(),
+            );
+        }
+
+        $addProductsMiddleware = $this->createAddProductsMiddleware($quantifiedItemPrices);
+
+        $result = $addProductsMiddleware->handle($orderProcessingData, $this->createOrderProcessingStack());
+        $actualOrderData = $result->orderData;
+
+        $this->assertThat(
+            $actualOrderData->getTotalPriceForItemTypes([OrderItemTypeEnum::TYPE_PRODUCT]),
+            new IsPriceEqual($expectedTotalPrice),
+        );
+
+        $this->assertThat(
+            $actualOrderData->totalPrice,
+            new IsPriceEqual($expectedTotalPrice),
+        );
+
+        $actualProductItemsType = $actualOrderData->getItemsByType(OrderItemTypeEnum::TYPE_PRODUCT);
+
+        $this->assertCount(count($products), $actualProductItemsType);
+        $this->assertCount(count($products), $actualOrderData->items);
+
+        foreach ($products as $index => $product) {
+            $this->assertSame($product->getName('en'), $actualProductItemsType[$index]->name);
+            $this->assertSame($product, $actualProductItemsType[$index]->product);
+        }
+    }
+
+    /**
+     * @return iterable
+     */
+    public function productsProvider(): iterable
+    {
+        yield 'single product' => [
+            'productsProviderInput' => [
+                [
+                    'name' => 'product1',
+                    'quantity' => 1,
+                    'unitPrice' => new Price(Money::create('100'), Money::create('121')),
+                ],
+            ],
+            'expectedTotalPrice' => new Price(Money::create('100'), Money::create('121')),
+        ];
+
+        yield 'multiple products' => [
+            'productsProviderInput' => [
+                [
+                    'name' => 'product1',
+                    'quantity' => 1,
+                    'unitPrice' => new Price(Money::create('100'), Money::create('121')),
+                ],
+                [
+                    'name' => 'product2',
+                    'quantity' => 1,
+                    'unitPrice' => new Price(Money::create('1000'), Money::create('1210')),
+                ],
+            ],
+            'expectedTotalPrice' => new Price(Money::create('1100'), Money::create('1331')),
+        ];
+
+        yield 'various quantities' => [
```

Comment:
```suggestion
        yield 'multiple products with various quantities' => [
```

nitpick :slightly_smiling_face: